src/storage-operators/src/s3_oneshot_sink.rs

```diff
@@ -56,33 +61,301 @@ pub fn copy_to<G, F>(
+/// initialization operation.
+fn render_initialization_operator<G>(
+    scope: G,
+    connection_context: ConnectionContext,
+    aws_connection: AwsConnection,
+    connection_id: GlobalId,
+    sink_id: GlobalId,
+    s3_key_manager: S3KeyManager,
+) -> Stream<G, Result<(), String>>
+where
+    G: Scope<Timestamp = Timestamp>,
+{
+    let worker_id = scope.index();
+    let num_workers = scope.peers();
+    let leader_id = usize::cast_from((sink_id, "initialization").hashed()) % num_workers;
+    let is_leader = worker_id == leader_id;
+
+    let mut builder =
+        AsyncOperatorBuilder::new("CopyToS3-initialization".to_string(), scope.clone());
+
+    let (mut start_handle, start_stream) = builder.new_output();
+
+    builder.build(move |caps| async move {
+        let [start_cap] = caps.try_into().unwrap();
+
+        // fallible async block to use the `?` operator for convenience
+        let leader_work = || async move {
+            info!(%sink_id, %worker_id, "s3 leader worker initialization");
+            let sdk_config = aws_connection
+                .load_sdk_config(&connection_context, connection_id, InTask::Yes)
+                .await?;
+
+            let client = mz_aws_util::s3::new_client(&sdk_config);
+
+            // Check that the S3 bucket path is empty before beginning the upload,
+            // and upload the INCOMPLETE sentinel file to the S3 path.
+            // Since we race against other replicas running the same sink we allow
+            // for objects to exist in the path if they were created by this sink
+            // (identified by the sink_id prefix).
+            // TODO: Add logic to determine if other replicas have already completed
+            // a full upload, to avoid writing the incomplete sentinel file again
+            // if the upload is already done.
+            mz_ore::task::spawn(|| "copytos3:initialization", {
+                let client = client.clone();
+                let bucket = s3_key_manager.bucket.clone();
+                let path_prefix = s3_key_manager.path_prefix().to_string();
+                let incomplete_sentinel_key = s3_key_manager.incomplete_sentinel_key();
+                async move {
+                    match mz_aws_util::s3::list_bucket_path(&client, &bucket, &path_prefix).await {
+                        Ok(Some(files)) => {
+                            let files = files
+                                .iter()
+                                .filter(|key| !s3_key_manager.is_sink_object(key))
+                                .collect::<Vec<_>>();
+                            if !files.is_empty() {
+                                Err(anyhow::anyhow!(
+                                    "S3 bucket path is not empty, contains: {:?}",
+                                    files
+                                ))?;
+                            }
+                        }
+                        Ok(None) => {}
+                        Err(e) => Err(e)?,
```

Comment:
Should this `?` move in `mz_aws_util::s3::list_bucket_path(&client, &bucket, &path_prefix).await?` and only match on the `Option<_>`?